File: app/src/main/java/br/com/concrete/desafio/until/RepositoryAdapter.java
package br.com.concrete.desafio.until;

import android.content.Context;
import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.constraint.ConstraintLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;

import br.com.concrete.desafio.PullRequestActivity;
import br.com.concrete.desafio.R;
import br.com.concrete.desafio.model.Repository;

public class RepositoryAdapter extends GenericAdapter<Repository> {

    public RepositoryAdapter(Repository[] dataSet) {
        super(dataSet);
    }

    @NonNull
    @Override
    public GenericViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        View view = LayoutInflater
                .from(viewGroup.getContext())
                .inflate(R.layout.rv_element_repository, viewGroup, false);

        return new GenericViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final GenericViewHolder viewHolder, int i) {
        final ConstraintLayout constraintLayout = viewHolder.constraintLayout;
        final Context context = constraintLayout.getContext();

        final TextView tvOwner = constraintLayout.findViewById(R.id.tvOwner);
        final TextView tvName = constraintLayout.findViewById(R.id.tvName);
        final TextView tvDescription = constraintLayout.findViewById(R.id.tvDescription);
        final TextView tvStars = constraintLayout.findViewById(R.id.tvStars);
        final TextView tvForks = constraintLayout.findViewById(R.id.tvForks);
        final ImageView ivPhoto = constraintLayout.findViewById(R.id.ivPhoto);

        if (dataSet.size() == 0) return;

        tvOwner.setText(dataSet.get(i).getOwner().getLogin());
        tvName.setText(dataSet.get(i).getName());
        tvStars.setText(String.valueOf(dataSet.get(i).getStargazers_count()));
        tvForks.setText(String.valueOf(dataSet.get(i).getForks_count()));

        String description = dataSet.get(i).getDescription();
        if (description.length() > 0) tvDescription.setText(description);
        else tvDescription.setText(context.getString(R.string.no_description));

        String photoUrl = dataSet.get(i).getOwner().getAvatar_url() + "&s=100";
        Picasso.get().load(photoUrl).into(ivPhoto);

        constraintLayout.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String pullUrl = dataSet.get(viewHolder.getAdapterPosition()).getPulls_url();
                pullUrl = pullUrl.substring(0, pullUrl.lastIndexOf("/pulls") + 1);

                String name = dataSet.get(viewHolder.getAdapterPosition()).getName();

                Intent intent = new Intent(context, PullRequestActivity.class);
                intent.putExtra("NAME", name);
                intent.putExtra("URL", pullUrl);
                context.startActivity(intent);
            }
        });
    }
}

File: app/src/main/java/br/com/renan/desafioandroid/repository/view/RepositoryAdapter.kt
package br.com.renan.desafioandroid.repository.view

import android.content.Intent
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import br.com.renan.desafioandroid.R
import br.com.renan.desafioandroid.model.data.Repository
import br.com.renan.desafioandroid.pullrequest.view.PullRequestActivity
import com.bumptech.glide.Glide
import kotlinx.android.synthetic.main.item_repository.view.*

class RepositoryAdapter(repositoryItemsList: List<Repository>) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private val recyclerList: List<Repository> = repositoryItemsList

    override fun onCreateViewHolder(parent: ViewGroup, p1: Int): RecyclerView.ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_repository, parent, false)
        )
    }

    override fun getItemCount(): Int {
        return recyclerList.size
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        (holder as ViewHolder).bind(recyclerList[position])
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(repository: Repository) = with(itemView) {

            Glide.with(itemView.context)
                .load(repository.owner.avatarUrl)
                .into(ivRepositoryAvatar)

            tvRepositoryName.text = repository.name
            tvRepositoryDescription.text = repository.description
            tvRepositoryForks.text = repository.forks.toString()
            tvRepositoryStars.text = repository.starts.toString()
            tvRepositoryUserName.text = repository.owner.login
            tvRepositoryFullName.text = repository.fullName.replace("/", " ")

            itemView.setOnClickListener { itemView ->
                val intent = Intent(itemView.context, PullRequestActivity::class.java)
                intent.putExtra("creator", repository.owner.login)
                intent.putExtra("repository", repository.name)
                context.startActivity(intent)
            }
        }
    }
}